### PR TITLE
Feat: Rename imports

### PR DESCRIPTION
**Pull request by predragnikolic**
When we rename or move a file,
the user is prompted to  rename all imports.

This PR also adds a new setting `updateImportsOnFileMove` with the following options `"prompt", "always", "never"`.

output

### CODE REVIEW

**Comment by predragnikolic** [comment]
Notes:
- I haven't tested this on Mac/Windows.
- This feature will work only when we rename one file. There are some File Managers that allows renaming multiple files at once,
 but this feature wont work in cases like that.

File: rename_imports.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            ignores=['**/node_modules/**'],
+            handler=HandleRenameImport
+        )
+
+
+queued_events = []  # type: List[FileWatcherEvent]
+
+
+class HandleRenameImport(FileWatcherProtocol):
+    def on_file_event_async(events: List[FileWatcherEvent]) -> None:
```
Comment (predragnikolic) [inline rename_imports.py:24]: The feature in this PR works...
but this line here bothers me. 
Notice that there is not `self` in the arguments.

Here is the interface for FileWatcherProtocol:
```
class FileWatcherProtocol(Protocol):
    def on_file_event_async(self, events: List[FileWatcherEvent]) -> None:
        """
        Called on file watcher events.
        This API must be triggered on async thread.

        :param events: The list of events to notify about.
        """
        ...
```

I tried to put `self`, but then I got runtime exceptions.

File: rename_imports.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+def register_rename_import_file_watcher(root_path: str):
+    file_watcher = get_file_watcher_implementation()
+    if file_watcher:
+        watcher = file_watcher.create(
+            root_path=root_path,
+            patterns=["**/*.ts", "**/*.tsx", "**/*.js", "**/*.jsx"],
+            events=["create", "delete"],
+            ignores=['**/node_modules/**'],
+            handler=HandleRenameImport
+        )
```
Comment (rchl) [inline rename_imports.py:17]: "create" returns a controller and it is responsibility of the callee to destroy() it when the server process is closing. Not doing that will result in the file watcher process running even after server has closed.

File: rename_imports.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+def register_rename_import_file_watcher(root_path: str):
+    file_watcher = get_file_watcher_implementation()
+    if file_watcher:
+        watcher = file_watcher.create(
+            root_path=root_path,
+            patterns=["**/*.ts", "**/*.tsx", "**/*.js", "**/*.jsx"],
+            events=["create", "delete"],
+            ignores=['**/node_modules/**'],
+            handler=HandleRenameImport
```
Comment (rchl) [inline rename_imports.py:16]: You need to pass an initialized instance.

File: rename_imports.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        )
+
+
+queued_events = []  # type: List[FileWatcherEvent]
```
Comment (rchl) [inline rename_imports.py:20]: Events should be bound to specific server instance, not global. Too much room for error otherwise.

File: rename_imports.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class HandleRenameImport(FileWatcherProtocol):
+    def on_file_event_async(events: List[FileWatcherEvent]) -> None:
+        global queued_events
+        queued_events.append(events[0])
```
Comment (rchl) [inline rename_imports.py:26]: Why is this appending only the first event?

File: rename_imports.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        sublime.set_timeout(check_if_there_was_a_rename, 10)
```
Comment (rchl) [inline rename_imports.py:27]: Doesn't seem like a reliable approach to have a timeout and check events from a different thread. This can result in race conditions and missing the two events you are looking for if there are more events besides that.

File: rename_imports.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def check_if_there_was_a_rename():
+    global queued_events
+    if len(queued_events) == 2:
+        first_event, new_name = queued_events[0]
+        second_event, old_name = queued_events[1]
+        if (first_event == 'create' and second_event == 'delete'):
```
Comment (rchl) [inline rename_imports.py:35]: This order is not guaranteed. I believe I got an opposite on Mac when testing this.

File: rename_imports.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            # Rename detected
+            prompt_rename(old_name, new_name)
+    queued_events = []
+
+
+def prompt_rename(old_name: str, new_name: str) -> None:
+    settings = sublime.load_settings("LSP-typescript.sublime-settings")
+    update_imports_on_file_move_setting = settings.get('settings', {}).get('updateImportsOnFileMove', "prompt")
+    if update_imports_on_file_move_setting == "never":
+        return
```
Comment (rchl) [inline rename_imports.py:45]: This could be checked before we even set up the watcher. Why set it up otherwise?

File: rename_imports.py (hunk 7 above)
Comment (rchl) [inline rename_imports.py:35]: I'm afraid that there is not really a reliable way to detect renames. It could be just two unrelated files being added and removed (for example when changing branches).

File: rename_imports.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    if update_imports_on_file_move_setting == 'prompt':
+        items = [
+            sublime.QuickPanelItem('Yes', 'Update imports.'),
+            sublime.QuickPanelItem('No', 'Do not update imports.')
+        ]
+        old_base_file_name = os.path.basename(old_name)
+        placeholder = "Update import paths for '{}'?".format(old_base_file_name)
+
+        def on_select(index):
+            global confirm_rename
```
Comment (rchl) [inline rename_imports.py:69]: unused

File: plugin.py [hunk 10]
```diff
@@ -54,6 +57,18 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
+    @classmethod
+    def is_allowed_to_start(
+        cls,
+        window: sublime.Window,
+        initiating_view: Optional[sublime.View] = None,
+        workspace_folders: Optional[List[WorkspaceFolder]] = None,
+        configuration: Optional[ClientConfig] = None
+    ) -> Optional[str]:
+        if workspace_folders:
+            register_rename_import_file_watcher(workspace_folders[0].path)
```
Comment (rchl) [inline plugin.py:69]: Should be done for every folder?

**Review by rchl — CHANGES_REQUESTED** [review]
Also a bunch of pyright and mypy issues in those changes.

File: rename_imports.py (hunk 3 above)
Comment (predragnikolic) [reply rename_imports.py:16]: I tired that, but got a `LSP-file-watcher-chokidar: ERROR: on_payload(): Handler already deleted`

Comment (predragnikolic) [reply rename_imports.py:16]: based on the code, I concluded that I should not pass an instance

LSP-file-watcher-chokidar will create the instance:
https://github.com/sublimelsp/LSP-file-watcher-chokidar/blob/main/watcher.py#L230

Comment (rchl) [reply rename_imports.py:16]: It doesn't create an instance. It's a weak reference to an instance so you need to call it to get the reference.

It's being deleted immediately, that's why you are getting "Handler already deleted". You need to store a reference to it somewhere, otherwise it will be GC-ed. That's also related to my other comment about having to bound the controller to the running server instance.

File: rename_imports.py (hunk 2 above)
Comment (predragnikolic) [reply rename_imports.py:17]: Do you know a hook to listen when the session is shutdown?
I could not find such method. 

I tried adding a `notification_handler` and listen to `exit` , but it is not called.

```
    @notification_handler("exit")
    def on_close(self, x):
        print("shutdown called")
```

Comment (rchl) [reply rename_imports.py:17]: `AbstractPlugin.on_session_end_async`

File: rename_imports.py (hunk 9 above)
Comment (predragnikolic) [reply rename_imports.py]: removed

File: rename_imports.py (hunk 7 above)
Comment (predragnikolic) [reply rename_imports.py]: I thought that chokidar has a way to detect a rename, but that is not the case.
there is a different lib, that supports rename detection => https://github.com/paulmillr/chokidar/issues/303#issuecomment-903710306

For the time being, I would rely on this logic:
```
        if len(self.events) == 2:
            create_event = next((event for event in self.events if event[0] == 'create'), None)
            delete_event = next((event for event in self.events if event[0] == 'delete'), None)
            if (create_event and delete_event):
                # Rename detected
                new_name = create_event[1]
                old_name = delete_event[1]
                self.prompt_rename(old_name, new_name)
```
If we have 2 events, 
and when we have a create and delete event,
only then prompt the rename....

>  It could be just two unrelated files being added and removed (for example when changing branches).

While than can happen, 
I do not know how to better to handle that than at the moment.

File: rename_imports.py (hunk 8 above)
Comment (predragnikolic) [reply rename_imports.py]: Changed the logic so that the watcher wont be created if the `update_imports_on_file_move_setting` is `never`

File: rename_imports.py (hunk 7 above)
Comment (predragnikolic) [reply rename_imports.py]: Fixed with explicitly finding those events:
```
create_event = next((event for event in self.events if event[0] == 'create'), None)
delete_event = next((event for event in self.events if event[0] == 'delete'), None)
if (create_event and delete_event):
```

File: rename_imports.py (hunk 5 above)
Comment (predragnikolic) [reply rename_imports.py]: I chanted it to `self.events.extend(events)`.

But here is the reason why I did used just the first event :arrow_down: 

<details>

When doing a rename on Linux,
the on_file_event_async is called twice.

Once for the create event, with the `events` argument being an array of one object.
And the second time, immediately after the first with the `events` argument being an array of one object.


Here is a better visual representation of the described behavior above:
```
# Rename a file
# on_file_event_async called
with events being [('create', '/home/predragnikolic/Documents/sandbox/my-react-ts-app/src/comp/Apaa.tsx')]
# on_file_event_async called
with events being [('delete', '/home/predragnikolic/Documents/sandbox/my-react-ts-app/src/comp/Apa.tsx')]
```

Initially I expected the `on_file_event_async` to be called only once, 
with the `events` argument being an array of **two** objects, but that is not the case.
</details>

File: rename_imports.py (hunk 4 above)
Comment (predragnikolic) [reply rename_imports.py]: moved to property in a class.

File: rename_imports.py (hunk 6 above)
Comment (predragnikolic) [reply rename_imports.py]: I change it to set_timeout_async :)

File: plugin.py [hunk 11]
```diff
@@ -54,6 +59,27 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
+        return None
+
+    def register_rename_import_file_watcher(self, window: sublime.Window, root_path: str) -> Optional[FileWatcher]:
+        settings = sublime.load_settings("LSP-typescript.sublime-settings")
+        update_imports_on_file_move_setting = settings.get('settings', {}).get('updateImportsOnFileMove', "prompt")
+        if update_imports_on_file_move_setting == "never":
+            return
+        file_watcher = get_file_watcher_implementation()
+        if file_watcher:
+            self.rename_import_handler = HandleRenameImport(window, root_path, file_watcher)
```
Comment (rchl) [inline plugin.py:81]: You are calling this method from a class method so `self` is in fact a class object (`cls`) and not an instance. You can't just set it on a class object as that will overwrite previously created instance.

Seems like the best (only) place would be `__init__` and getting the workspace folders from `Session`.

File: rename_imports.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+from LSP.plugin.core.file_watcher import FileWatcher
+from LSP.plugin.core.file_watcher import FileWatcherEvent
+from LSP.plugin.core.file_watcher import FileWatcherProtocol
+from LSP.plugin.core.typing import List, Type
+import os
+import sublime
+
+
+class HandleRenameImport(FileWatcherProtocol):
```
Comment (rchl) [inline rename_imports.py:9]: Weird name for a class. Would suggest `RenameImportHandler` for example.

File: rename_imports.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+    def on_file_event_async(self, events: List[FileWatcherEvent]) -> None:
+        self.events.extend(events)
+        sublime.set_timeout_async(self._check_if_there_was_a_rename, 10)
+
+    def _check_if_there_was_a_rename(self):
+        if len(self.events) == 2:
+            create_event = next((event for event in self.events if event[0] == 'create'), None)
+            delete_event = next((event for event in self.events if event[0] == 'delete'), None)
+            if (create_event and delete_event):
```
Comment (rchl) [inline rename_imports.py:32]: nit: no need for parenthesis

File: rename_imports.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        self.file_watcher.destroy()
+
+    def on_file_event_async(self, events: List[FileWatcherEvent]) -> None:
+        self.events.extend(events)
+        sublime.set_timeout_async(self._check_if_there_was_a_rename, 10)
+
+    def _check_if_there_was_a_rename(self):
+        if len(self.events) == 2:
+            create_event = next((event for event in self.events if event[0] == 'create'), None)
+            delete_event = next((event for event in self.events if event[0] == 'delete'), None)
```
Comment (rchl) [inline rename_imports.py:31]: This seems very fragile and implementation dependent. The logic should just be: if there are two sibling events where one is 'create' and one is 'delete' then trigger the rename.

Same for the timeout that calls this method. There is no guarantee that both events will come within 10ms and if they don't, we'll clear the list and miss the whole rename.

File: rename_imports.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                # Rename detected
+                new_name = create_event[1]
+                old_name = delete_event[1]
+                self.prompt_rename(old_name, new_name)
+        self.events = []
+
+    def prompt_rename(self, old_name: str, new_name: str) -> None:
+        settings = sublime.load_settings("LSP-typescript.sublime-settings")
+        update_imports_on_file_move_setting = settings.get('settings', {}).get('updateImportsOnFileMove', "prompt")
+        if update_imports_on_file_move_setting == "always":
```
Comment (rchl) [inline rename_imports.py:42]: Given the the logic for checking if the file was renamed can trigger false positives pretty easily, I think that there should be no "always" option as it seems dangerous. It should be a boolean option where true triggers the prompt.

And we should communicate both the old and new name ideally.

File: plugin.py [hunk 16]
```diff
@@ -54,6 +59,27 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
+        workspace_folders: Optional[List[WorkspaceFolder]] = None,
+        configuration: Optional[ClientConfig] = None
+    ) -> Optional[str]:
+        if workspace_folders:
+            cls.register_rename_import_file_watcher(cls, window, workspace_folders[0].path)
+        return None
+
+    def register_rename_import_file_watcher(self, window: sublime.Window, root_path: str) -> Optional[FileWatcher]:
+        settings = sublime.load_settings("LSP-typescript.sublime-settings")
+        update_imports_on_file_move_setting = settings.get('settings', {}).get('updateImportsOnFileMove', "prompt")
```
Comment (rchl) [inline plugin.py:76]: You have access to `configuration` so no need to read the settings manually.

**Review by rchl — CHANGES_REQUESTED** [review]

File: plugin.py [hunk 17]
```diff
@@ -54,6 +59,27 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._api = None  # type: Optional[ApiWrapperInterface]
```
[the remark this answers is not in the document]
Comment (rchl) [reply plugin.py:59]: That's because of the `self`/`cls` confusion.

File: rename_imports.py (hunk 7 above)
Comment (rchl) [reply rename_imports.py]: > I thought that chokidar has a way to detect a rename, but that is not the case.
> there is a different lib, that supports rename detection => [paulmillr/chokidar#303 (comment)](https://github.com/paulmillr/chokidar/issues/303#issuecomment-903710306)

Interesting watcher. At least if we are to trust the author's claims.

I wouldn't want to switch to using it as that's too much work and the claims would have to be tested in practice also but we could take inspiration from it and implement support for rename events in our watcher.

> I do not know how to better to handle that than at the moment.

Me neither that's why I'm suggesting to remove the potentially dangerous option to rename without confirmation.

Comment (rchl) [reply rename_imports.py]: > there is a different lib, that supports rename detection => [paulmillr/chokidar#303 (comment)](https://github.com/paulmillr/chokidar/issues/303#issuecomment-903710306)

BTW. The code style of that project is... interesting.
 - Blank lines between every line of code
 - excessive use of spaces (leading/trailing spaces within parenthesis for example)

https://github.com/fabiospampinato/watcher/blob/master/src/watcher.ts

It would not be fun to contribute to such project :)

File: rename_imports.py (hunk 5 above)
Comment (rchl) [reply rename_imports.py]: The problem with the initial approach was that it was based on observed implementation details which is not a good idea.

File: rename_imports.py (hunk 14 above)
Comment (rchl) [reply rename_imports.py:31]: And to clarify, I meant without checking if there are exactly 2 events as that is not guaranteed.

File: rename_imports.py (hunk 15 above)
Comment (predragnikolic) [reply rename_imports.py:42]: Agree, 
I will remove the setting in favor of a boolean option.
I will use the existing ts setting for this `allowRenameOfImportPath?: boolean;`.
https://github.com/armanio123/TypeScript/blob/fbf840114ef38aa044c6b4840d5a43747bcf8a4e/src/server/protocol.ts#L3392

File: plugin.py (hunk 10 above)
Comment (predragnikolic) [reply plugin.py]: I plan to add support form multiple folders, 
I just need more time to think about how to handle the cleanup logic (when one folder closes, and so on)

I was thinking of having a Map, where the key is the `root_path` and the value is a `RenameImportHandler` instance.

Comment (rchl) [reply plugin.py]: How would you know if the folders have changed?

File: rename_imports.py (hunk 14 above)
Comment (predragnikolic) [reply rename_imports.py:31]: > The logic should just be: if there are two sibling events where one is 'create' and one is 'delete' then trigger the rename.

We cannot rely on sibling events. :)

Here are the logs when renaming multiple files at once on Linux,  .
the folder structure is the following:
```
├── test
│   ├── file1.ts
│   ├── file2.ts
│   └── file3.ts
```

if we rename multiple files at once,
we will have the following events logged:
```
events [('delete', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file1.ts')]
events [('delete', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file2.ts')]
events [('delete', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file3.ts')]
events [('create', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file11.ts')]
events [('create', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file21.ts')]
events [('create', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file31.ts')]
```
<details>
<summary>
see the gif
</summary>

output

</details>

So to jump to my point is, 
we cannot rely on two sibilling events, 
as that would result in incorrect behavior. 

In the above example, it would mean to rename `file3.ts` to `file11.ts`, which is not correct.
```
events [('delete', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file3.ts')]
events [('create', '/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript/test/file11.ts')]
```

Comment (rchl) [reply rename_imports.py:31]: In that case it seems like we need more sophisticated solution. Something similar to what is mentioned in the link you've posted before (https://github.com/paulmillr/chokidar/issues/303#issuecomment-903710306)

Comment (rchl) [reply rename_imports.py:31]: I also can imagine that 3 renames in a row would not be handled well by the quick panel that we are triggering.

File: plugin.py (hunk 10 above)
Comment (predragnikolic) [reply plugin.py]: Multiple folders are semi supported.

The current code already supports sublime project.
When having a ST project with multiple folders:
```
// example.sublime-project
{
	"folders":
	[
		{
			"path": "/home/predragnikolic/Documents/sandbox/my-react-ts-app"
		},
		{
			"path": "/home/predragnikolic/Documents/sandbox/coc-tsserver"
		},
		{
			"path": "/home/predragnikolic/Documents/sandbox/volar"
		},
		{
			"path": "/home/predragnikolic/.config/sublime-text/Packages/LSP-typescript"
		}
	]
}

```

It works until the moment you try to add/remove folders to a project while ST is open. :)

File: rename_imports.py (hunk 14 above)
Comment (predragnikolic) [reply rename_imports.py:31]: I intentionally didn't want to support multiple file renames at once. only because I do not want to spam the user will prompts.

I only wanted to cover the case when a user renames a single file.

Comment (predragnikolic) [reply rename_imports.py:31]: > In that case it seems like we need more sophisticated solution. Something similar to what is mentioned in the link you've posted before ...

Will improve the logic by checkin the `iNode` number of a file. :)

Comment (predragnikolic) [reply rename_imports.py:31]: well,
I tried to use `os.stat(file_name).st_ino` to get the ino number for a file.

But I quickly run into problems.

Here was my soloution:

image

```
events_with_ino = [(event[0], event[1], os.stat(event[1]).st_ino) for event in events]
#					^^^^^^^^ event_type
#							  ^^^^^^^^ file_name
#										^^^^^^^^^^^^^^^^^^^^^^^^ ino number for that file_name
```
I tried to attach an ino number for a file_name,
but this works only for files that exist.

That said os.stat(event[1]).st_ino will work only for `create` or `change` events,
but not for when the event type is `delete`.
Because by the time we receive the delete event in python,
the file no longer exists,
so calling `os.stat(file_name_that_no_longer_exist).st_ino` will throw.  

----

Ideally I would shift this logic to be the responsibility of the file_watcher, 
so we have an interface like this:
https://github.com/neoclide/coc-tsserver/blob/master/src/server/features/updatePathOnRename.ts#L33

File: plugin.py [hunk 18]
```diff
@@ -52,10 +55,24 @@ class LspTypescriptPlugin(NpmClientHandler):
 
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._api = None  # type: Optional[ApiWrapperInterface]
+        self.rename_import_handler = None  # type: Optional[RenameImportHandler]
         super().__init__(*args, **kwargs)
 
+    def register_rename_import_file_watcher(self, session: Session) -> None:
+        allow_rename_of_import_path = session.config.init_options.get('preferences.allowRenameOfImportPath')
+        workspace_folders = session.get_workspace_folders()
```
Comment (rchl) [inline plugin.py:63]: This variable is only needed with the `if file_watcher` condition.

File: plugin.py (hunk 10 above)
Comment (rchl) [reply plugin.py]: The current code only takes the path of the first folder.

File: rename_imports.py (hunk 14 above)
Comment (rchl) [reply rename_imports.py:31]: Supporting this in the watcher is the way to go but not sure if we would be able to do it in `LSP-file-watcher-chokidar`. Might need to be implemented upstream in `chokidar` instead.

**Comment by rchl** [comment]
Quite a lot of work was put into it but ultimately I don't think that this is the way to go. It would not be fine to trigger disrupting rename prompts whenever something would be renamed. Renames should only be handled by the server when user explicitly asks to rename some file so we need LSP-specific rename actions instead.

Maybe if there would be a way to offer the user suggestion to update imports and non-distracting way then it would be more acceptable but it might be better to just only do that on explicit request to rename.

**Comment by predragnikolic** [comment]
Agreed, I will close this PR.

I saw the willRename PR on the ts server repo.
I believe that is the way to go.
